qiskit/synthesis/evolution/product_formula.py

```diff
@@ -177,6 +186,64 @@ def evolve_pauli(
         _multi_qubit_evolution(output, pauli, time, cx_structure, wrap)
 
 
+def reorder_paulis(
+    operators: SparsePauliOp | list[SparsePauliOp],
+) -> SparsePauliOp | list[SparsePauliOp]:
+    r"""
+    Creates an equivalent operator by reordering terms in order to yield a
+    shallower circuit after evolution synthesis. The original operator remains
+    unchanged.
+
+    This method works in three steps. First, a graph is constructed, where the
+    nodes are the terms of the operator and where two nodes are connected if
+    their term acts on the same qubit (for example, the terms $IXX$ and $IYI$
+    would be connected, but not $IXX$ and $YII$). Then, the graph is colored.
+    Two terms with the same color thus do not act on the same qubit, and in
+    particular, their evolution subcircuits can be run in parallel in the
+    greater evolution circuit of ``operator``. Finally, a new
+    :class:`~qiskit.quantum_info.SparsePauliOp` is created where terms of the
+    same color are grouped together.
+
+    If the input is in fact a list of
+    :class:`~qiskit.quantum_info.SparsePauliOp`, then the terms of all operators
+    will be coalesced and reordered into a single
+    :class:`~qiskit.quantum_info.SparsePauliOp`.
+
+    Args:
+        operators: The operator or list of operators whose terms to reorder.
+    """
+    if not isinstance(operators, list):
+        operators = [operators]
+    # Do nothing in trivial cases
+    if not (
+        # operators is a list of > 1 SparsePauliOp
+        len(operators) > 1
+        # operators has a single SparsePauliOp, which has > 1 terms
+        or (len(operators) == 1 and len(operators[0]) > 1)
+    ):
+        return operators
+    graph = rx.PyGraph()
+    graph.add_nodes_from(sum((o.to_list() for o in operators), []))
+    indexed_nodes = list(enumerate(graph.nodes()))
+    for (idx1, term1), (idx2, term2) in combinations(indexed_nodes, 2):
+        # Add an edge between term1 and term2 if they touch the same qubit
+        for a, b in zip(term1[0], term2[0]):
+            if not (a == "I" or b == "I"):
+                graph.add_edge(idx1, idx2, None)
+                break
+    # TODO: The graph is likely not very big, so an exact coloring could be
+    # computed in a reasonable time.
+    coloring = rx.graph_greedy_color(graph)
```

Comment:
A quick comment: the Rustworkx's `graph_greedy_color` function can take an optional argument `strategy` that specifies the heuristic coloring method used, please see https://www.rustworkx.org/dev/apiref/rustworkx.ColoringStrategy.html for detail. You might want to experiment with the different strategies available to see which performs best for this application (or run all three one after another and choose the result with the smallest number of colors).

Reply:
I agree, we should try this or expose an argument allowing to set the coloring strategy.